File: kirun-js/src/engine/function/system/array/Join.ts
import { Schema } from '../../../json/schema/Schema';
import { SchemaType } from '../../../json/schema/type/SchemaType';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { AbstractFunction } from '../../AbstractFunction';

const VALUE = 'source';
const DELIMITTER = 'delimiter';
const OUTPUT = 'result';

const SIGNATURE = new FunctionSignature('Join')
    .setNamespace(Namespaces.SYSTEM_ARRAY)
    .setParameters(
        new Map([
            [
                VALUE,
                new Parameter(
                    VALUE,
                    Schema.ofArray(
                        VALUE,
                        Schema.of(
                            'each',
                            SchemaType.STRING,
                            SchemaType.INTEGER,
                            SchemaType.LONG,
                            SchemaType.DOUBLE,
                            SchemaType.FLOAT,
                            SchemaType.NULL,
                        ),
                    ),
                ),
            ],
            [
                DELIMITTER,
                new Parameter(DELIMITTER, Schema.ofString(DELIMITTER).setDefaultValue('')),
            ],
        ]),
    )
    .setEvents(new Map([Event.outputEventMapEntry(new Map([[OUTPUT, Schema.ofString(OUTPUT)]]))]));

export class Join extends AbstractFunction {
    public getSignature(): FunctionSignature {
        return SIGNATURE;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let source: any[] = context?.getArguments()?.get(VALUE);
        let delimitter: string = context?.getArguments()?.get(DELIMITTER);

        return new FunctionOutput([
            EventResult.outputOf(new Map([[OUTPUT, source.join(delimitter)]])),
        ]);
    }
}

File: kirun-py/src/kirun_py/function/system/array/join.py
from __future__ import annotations

from typing import Any, List, TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.json.schema.type.schema_type import SchemaType
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

VALUE = 'source'
DELIMITER = 'delimiter'
OUTPUT = 'result'


class Join(AbstractFunction):

    def __init__(self) -> None:
        super().__init__()
        self._signature = (
            FunctionSignature('Join')
            .set_namespace(Namespaces.SYSTEM_ARRAY)
            .set_parameters({
                VALUE: Parameter(
                    VALUE,
                    Schema.of_array(
                        VALUE,
                        Schema.of(
                            'each',
                            SchemaType.STRING,
                            SchemaType.INTEGER,
                            SchemaType.LONG,
                            SchemaType.DOUBLE,
                            SchemaType.FLOAT,
                            SchemaType.NULL,
                        ),
                    ),
                ),
                DELIMITER: Parameter(
                    DELIMITER,
                    Schema.of_string(DELIMITER).set_default_value(''),
                ),
            })
            .set_events(dict([
                Event.output_event_map_entry(
                    {OUTPUT: Schema.of_string(OUTPUT)}
                )
            ]))
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        source: List[Any] = context.get_arguments().get(VALUE)
        delimiter: str = context.get_arguments().get(DELIMITER)

        result = delimiter.join(str(x) if x is not None else '' for x in source)

        return FunctionOutput([
            EventResult.output_of({OUTPUT: result})
        ])
